File: 0069-sqrtx/0069-sqrtx.cpp
class Solution {
public:
    int mySqrt(int x) {
        if (x < 2) return x;
        int l = 1;
        int r = x / 2;

        while (l <= r) {
            int m = r - (r - l) / 2;
            if (m > x / m) r = m - 1;
            else if (m < x / m) l = m + 1;
            else return m;
        }
        return r;      
    }
};

File: 0069-sqrtx/0069-sqrtx.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x < 2:
            return x
        
        l, r = 0, x // 2
        while l <= r:
            mid = (l + r) // 2
            if mid * mid < x:
                l = mid + 1
            elif mid * mid == x:
                return mid
            else:
                r = mid - 1
                
        return r
        